test: run mempool_expiry.py even with wallet disabled

mjdietzx opened the pull request:
Run the mempool expiry test even when the wallet was not compiled, as proposed in https://github.com/bitcoin/bitcoin/issues/20078.

MarcoFalke commented on test/functional/mempool_expiry.py:
```diff
@@ -27,54 +27,52 @@ class MempoolExpiryTest(BitcoinTestFramework):
-        child_txid = node.sendrawtransaction(child_signed)
+
+        # Broadcast the child transaction spending the parent transaction.
+        parent_utxo = wallet.get_utxo()
```
Pretty sure this is not the parent change. miniwallet should have two outputs at this point. One leftover coinbase and one change. Coin selection favors the largest output, so the coinbase one.

I might be wrong, but I think it could make sense to check that the child tx really has parent_txid as ancestor via the mempool rpc

mjdietzx replied:
I added an assertion verifying this to make it clear https://github.com/bitcoin/bitcoin/pull/20276/commits/de4173bacd3a2ce6294515072cb725d5c1074cd8.

Originally I had:
```python
parent_utxo = next(filter(lambda utxo: utxo['txid'] == parent_txid, wallet._utxos))
```

but then I realized `get_utxo()` would do this it me: ```"""Return the last utxo. Can be used to get the change output immediately after a send_self_transfer"""```

MarcoFalke replied:
Oh, sorry my bad in that case. Though, the mempool check would still make sense.

mjdietzx commented:
@MarcoFalke not to beat this to death, but I added another test case in https://github.com/bitcoin/bitcoin/pull/20276/commits/71081df19a816b5c243a6bcccab9fb11888732fa. While convincing myself that I was using `miniWallet` properly, and not getting lucky with any passing tests, I wrote this and figured it was useful enough to include in the PR

mjdietzx commented:
Rebased and should be ready to merge

0xB10C reviewed (commented):
ACK 124c8ab5df85b606e00cfa666de6105e0620de00

Thank you @mjdietzx for this PR! Linking my original PR that added this test for context:  https://github.com/bitcoin/bitcoin/pull/18172

mjdietzx commented on test/functional/mempool_expiry.py:
```diff
@@ -26,57 +26,62 @@
-    def skip_test_if_missing_module(self):
-        self.skip_if_no_wallet()
+    def sync_mempool(self):
+        """Expiry of mempool txs are only checked when a new tx is added to the mempool."""
```
Thanks for the suggestion, I made this change and squashed it into the commit

0xB10C commented:
re-ACK ec66561d87881f54d64c8d633e267efe1b18b4ad

Only changes are:
- https://github.com/bitcoin/bitcoin/pull/20276#discussion_r532085685 addressed
- formatting of long lines changed: removed line breaks (multiple lines -> one line)

mjdietzx commented:
So since we don't use the wallet in mempool_expiry.py after this PR, does this close https://github.com/bitcoin/bitcoin/issues/19448? Also, since https://github.com/bitcoin/bitcoin/pull/19450 is described as a temp workaround would we revert it after this is merged? cc @hebasto 

Can someone explain why (only?) this test is triggering a data race though? Everything seems pretty simple so I'm confused why only this test would expose it. And I wonder if it's not a hint at an underlying problem?

hebasto commented:
> So since we don't use the wallet in mempool_expiry.py after this PR, does this close #19448? Also, since #19450 is described as a temp workaround would we revert it after this is merged? cc @hebasto

As noted in #19448 OP other similar issues are reported: #19211 and #19417. I think we could just wait until Berkley DB will [no longer used](https://github.com/bitcoin/bitcoin/issues/20160) in Bitcoin Core.

> Can someone explain why (only?) this test is triggering a data race though? Everything seems pretty simple so I'm confused why only this test would expose it. And I wonder if it's not a hint at an underlying problem?

I believe something is wrong with Berkeley DB code, but did not check it myself.

MarcoFalke commented on test/functional/mempool_expiry.py:
```diff
@@ -26,73 +26,78 @@
+        when a new transaction is added to the mempool."""
+        try:
+            self.wallet.send_self_transfer(from_node=self.nodes[0])
+        except JSONRPCException:
```
can you explain why this is needed? I'd prefer not to call the method when it won't do anything anyway

0xB10C replied:
[`CTxMemPool::Expire(..)`](https://github.com/bitcoin/bitcoin/blob/cd720337fe9c27422fe8d507726ca8c79f52b4ba/src/txmempool.cpp#L945) is only called in [ `LimitMempoolSize(..)`](https://github.com/bitcoin/bitcoin/blob/f17e8ba3a17b6516a1b1fb7f45d506a339e99f90/src/validation.cpp#L331) which is only called in [`UpdateMempoolForReorg(..)`](https://github.com/bitcoin/bitcoin/blob/f17e8ba3a17b6516a1b1fb7f45d506a339e99f90/src/validation.cpp#L370) and [`MemPoolAccept::Finalize(..)`](https://github.com/bitcoin/bitcoin/blob/f17e8ba3a17b6516a1b1fb7f45d506a339e99f90/src/validation.cpp#L983). 

To trigger the expiry of a transaction in the mempool we send an unrelated transaction. For the unrelated transaction `MemPoolAccept::Finalize(..)` is called which calls `LimitMempoolSize(..)` which calls `CTxMemPool::Expire(..)` on all timed-out transactions. If we don't send a new transaction, we can't test a transaction is really removed.

MarcoFalke replied:
I meant the `except: pass` part, sorry for being unclear :sweat_smile:

mjdietzx replied:
There's a minor problem here, re the **unrelated transaction** part of @0xB10C's explanation. With `MiniWallet` if we don't do any extra bookkeeping in this test, and just do `self.wallet.send_self_transfer(from_node=node)` to trigger the mempool update, the (now expired) utxo will be spent. So this `send_self_transfer` will throw `test_framework.authproxy.JSONRPCException: mempool full (-26)`

We'd need to ensure that we don't spend one of the expired UTXOs when we want to `trigger_mempool_expiry` if we don't catch the Exception. ie:
```C++
node.setmocktime(expiry_time)
# Broadcasts an unrelated transaction as the expiry of the transactions in the mempool is only checked when a new transaction is added to the mempool.
new_utxo = self.wallet.get_utxo(txid=independent_txid)
new_txid = self.wallet.send_self_transfer(from_node=self.nodes[0], utxo_to_spend=new_utxo)['txid']
# parent_txid has now been evicted from the mempool.
```

Because the mempool is still updated if `send_self_transfer` fails, I thought it was cleaner to catch the Exception rather than have extra logic (that's unrelated to what we're testing) to ensure we're not spending expired UTXOs in this trigger.

Maybe I need to add a comment along these lines in the `except: pass` to make this clear? I'm realizing this is more confusing than I thought

MarcoFalke replied:
I don't understand why spending an non-existent coin will throw 'mempool full' (not missing-inputs) and how that is going to trigger mempool expiry of other txs.

Anyway, I guess the options are:

* Add a comment
* Remember the previous utxo (not expired) and use that
* Drop utxos that are about to expire from the wallet (might require fresh coins later on in the script)
* Something else?

mjdietzx replied:
So I figured out why there were some weird errors that didn't make sense as you point out. After the first subtest some UTXOs were hanging around in the mempool, that ended up being used in the second subtest because MiniWallet sorts for the largest UTXOs. That's why I ensure the mempool is clear at the beginning of each subtest run now.

MarcoFalke commented on test/functional/mempool_expiry.py:
```diff
@@ -26,73 +26,78 @@
-        # Let half of the timeout elapse and broadcast the child transaction.
-        half_expiry_time = entry_time + int(60 * 60 * timeout/2)
+        # Let half of the timeout elapse.
+        half_expiry_time = entry_time + 3600 * timeout // 2
```
it looks like you are fixing up style in the same commit as refactors, cleanups and miniwallet changes

It would be easier to follow if the formatting changes, the `trigger_mempool_expiry`, and the miniwallet changes were all in separate commits, where each passed on its own and can be reviewed on its own

mjdietzx replied:
Ah yeah, I did too much extra stuff in this PR. I'm gonna undo all this unrelated style/refactor stuff. Will undo the string formatting changes as well

MarcoFalke commented on test/functional/mempool_expiry.py:
```diff
@@ -26,73 +26,78 @@
     def run_test(self):
-        self.log.info('Test default mempool expiry timeout of %d hours.' %
-                      DEFAULT_MEMPOOL_EXPIRY)
+        self.log.info('Test default mempool expiry timeout of {} hours.'.format(DEFAULT_MEMPOOL_EXPIRY))
```
not sure about mixing all the formatting changes in with the other changes. In either case if you want to change this, you can use f-strings

maflcko reviewed (commented):
concept ACK, but the changes were a bit too much for me to properly review

mjdietzx committed:
test: run mempool_expiry.py even with wallet disabled

Test coverage is also extended in this commit to:
Ensure that another transaction in the mempool is not evicted
when other transactions expire. Note: this other transaction does
not have any ancestors in the mempool that expired. Otherwise it
would be evicted from the mempool, and we already test this case.

mjdietzx commented:
Alright, I cleaned this up and got rid of all the fluff that wasn't related to the core of this PR. Should be ready to go now, will make sure I keep PRs focused in the future

MarcoFalke commented on test/functional/mempool_expiry.py:
```diff
@@ -26,44 +26,48 @@
+        self.wallet = MiniWallet(node)
+
+        # Ensure the mempool starts empty.
+        node.generate(1)
```
Why is this needed? You are generating a block in the next line, which cleares the mempool

mjdietzx replied:
TLDR; I made everything explicit in this commit https://github.com/bitcoin/bitcoin/pull/20276/commits/4b99dc5ef62fe1d1a6e23d71f55570dc14aac9fa because I realized this is all confusing. Thanks for pointing this out. 

It was because fees from the transactions in the first test run (that are still in the mempool) `self.test_transaction_expiry(DEFAULT_MEMPOOL_EXPIRY)` get included in `MiniWallet`'s utxo set when this runs for the next sub-test `self.test_transaction_expiry(CUSTOM_MEMPOOL_EXPIRY)` if the first block is generated with `self.wallet.generate`

Then because of the way `MiniWallet` sorts UTXOs based on which is largest it means we end up using an expired transaction for the "trigger" transaction, and get a weird error in the 2nd test run, but not the first - which threw me for a loop

MarcoFalke commented on test/functional/mempool_expiry.py:
```diff
@@ -26,44 +26,48 @@
+        # Ensure the mempool starts empty.
+        node.generate(1)
+        assert_equal(set(node.getrawmempool()), set())
+        # Add enough mature utxos to the wallet so that all txs spend confirmed coins.
```
this comment is not true. The "trigger" txs are chained txs. More inputs are needed.

mjdietzx replied:
That's a good point, addressed in most recent commit

MarcoFalke commented:
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

mjdietzx commented:
This has been squashed/rebased and is ready to go

MarcoFalke commented:
ACK 3b064fcb9dd3df6c438a440f0fea86e9cf7b5f57
